lib/spack/llnl/util/filesystem.py

```diff
@@ -833,7 +848,24 @@ def names(self):
         Returns:
             list of strings: A list of library names
         """
-        return list(dedupe(x.split('.')[0][3:] for x in self.basenames))
+        names = []
+
+        for x in self.basenames:
+            name = x
+            if x.startswith('lib'):
+                name = x[3:]
+
+            # Valid extensions include: ['.dylib', '.so', '.a']
+            for ext in ['.dylib', '.so', '.a']:
+                i = name.rfind(ext)
+                if i != -1:
+                    names.append(name[:i])
+                    break
+            else:
+                # No valid extension, should we still include it?
```

Comment:
This should still work, as `rfind` will find a valid extension (`.so`). The `else:` here is for libraries that don't have `.dylib`, `.so`, or `.a` anywhere in the filename (which probably don't exist). It might not make much sense for LibraryList, but I think it makes sense for HeaderList. Eigen has header files with no extensions, for example.